### PR TITLE
Bug:2076270  better VM deletion handling

### PR DESCRIPTION
**Pull request by eslutsky**
this  PR fixes the VM deletion issues found after the refactor https://github.com/openshift/cluster-api-provider-ovirt/pull/131

### CODE REVIEW

File: pkg/actuators/machine/machine.go [hunk 1]
```diff
@@ -302,7 +302,22 @@ func (ms *machineScope) delete() error {
 		}
 		return errors.Wrap(err, "error finding VM by name")
 	}
-	return ms.ovirtClient.RemoveVM(vm.ID(), ovirtC.ContextStrategy(ms.Context))
+	if vm == nil {
```
Comment (janosdebugs) [inline pkg/actuators/machine/machine.go:305]: This case is already handled in line 300 above. Why is this check here again?

File: pkg/actuators/machine/machine.go [hunk 2]
```diff
@@ -302,7 +302,22 @@ func (ms *machineScope) delete() error {
...
+		klog.Infof("Skipped deleting a VM that is already deleted.\n")
+		return nil
+	}
+
+	if err := vm.Stop(true); err != nil {
```
Comment (janosdebugs) [inline pkg/actuators/machine/machine.go:310]: Please pass the context here.

File: pkg/actuators/machine/machine.go [hunk 3]
```diff
@@ -302,7 +302,22 @@ func (ms *machineScope) delete() error {
...
+		return err
+	}
+	if _, err := vm.WaitForStatus(ovirtC.VMStatusDown); err != nil {
```
Comment (janosdebugs) [inline pkg/actuators/machine/machine.go:313]: Please pass the context here.

File: pkg/actuators/machine/machine.go [hunk 4]
```diff
@@ -302,7 +302,22 @@ func (ms *machineScope) delete() error {
...
+		return err
+	}
+	if err := vm.Remove(); err != nil && !ovirtC.HasErrorCode(err, ovirtC.ENotFound) {
```
Comment (janosdebugs) [inline pkg/actuators/machine/machine.go:316]: Please pass the context here.

**Review by ghost — CHANGES_REQUESTED** [review]

File: pkg/actuators/machine/machine.go (hunk 1 above)
Comment (eslutsky) [reply pkg/actuators/machine/machine.go:305]: sure , with this PR https://github.com/oVirt/go-ovirt-client/pull/181 we wont need it, removing

File: pkg/actuators/machine/machine.go [hunk 5]
```diff
@@ -302,7 +302,17 @@ func (ms *machineScope) delete() error {
 		}
 		return errors.Wrap(err, "error finding VM by name")
 	}
-	return ms.ovirtClient.RemoveVM(vm.ID(), ovirtC.ContextStrategy(ms.Context))
+	if err := vm.Stop(true, ovirtC.ContextStrategy(ms.Context)); err != nil {
```
Comment (janosdebugs) [inline pkg/actuators/machine/machine.go:305]: What happens here if the VM is already stopped? If it would return an error, we should handle that.

**Review by ghost — APPROVED** [review]
/lgtm

File: pkg/actuators/machine/machine.go (hunk 2 above)
Comment (eslutsky) [reply pkg/actuators/machine/machine.go]: fixed

File: pkg/actuators/machine/machine.go (hunk 3 above)
Comment (eslutsky) [reply pkg/actuators/machine/machine.go]: fixed

File: pkg/actuators/machine/machine.go (hunk 4 above)
Comment (eslutsky) [reply pkg/actuators/machine/machine.go]: fixed

File: pkg/actuators/machine/machine.go (hunk 5 above)
Comment (eslutsky) [reply pkg/actuators/machine/machine.go:316]: we have tested this scenario, this won't trigger any errors.

**Comment by eslutsky** [comment]
the installation is not passing with this PR, workers are not created, and we need to find the root cause before we merge this.

**Review by ghost — APPROVED** [review]

File: pkg/controllers/nodeController/nodeController.go [hunk 6]
```diff
@@ -59,16 +59,16 @@ func (r *nodeController) Reconcile(ctx context.Context, request reconcile.Reques
 
 	vm, err := ovirtClient.GetVMByName(node.Name)
 	if err != nil {
+		if !ovirtC.HasErrorCode(err, ovirtC.ENotFound) {
```
Comment (janosdebugs) [inline pkg/controllers/nodeController/nodeController.go:62]: The other way around, no?

File: pkg/actuators/machine/machine.go [hunk 7]
```diff
@@ -297,11 +299,14 @@ func (ms *machineScope) create() error {
 
 // exists returns true if machine exists.
 func (ms *machineScope) exists() (bool, error) {
-	vm, err := ms.ovirtClient.GetVMByName(ms.machine.Name, ovirtC.ContextStrategy(ms.Context))
+	_, err := ms.ovirtClient.GetVMByName(ms.machine.Name, ovirtC.ContextStrategy(ms.Context))
 	if err != nil {
+		if !ovirtC.HasErrorCode(err, ovirtC.ENotFound) {
```
Comment (janosdebugs) [inline pkg/actuators/machine/machine.go:304]: The other way around.

File: pkg/controllers/providerIDcontroller/providerIDController.go [hunk 8]
```diff
@@ -66,19 +67,20 @@ func (r *providerIDController) Reconcile(ctx context.Context, request reconcile.
...
-	ovirtC, err := r.GetoVirtClient()
+	ovirtclient, err := r.GetoVirtClient()
 	if err != nil {
 		return "", errors.Wrap(err, "error getting connection to oVirt")
 	}
 
-	vm, err := ovirtC.GetVMByName(nodeName)
+	vm, err := ovirtclient.GetVMByName(nodeName)
 	if err != nil {
+		if !ovirtC.HasErrorCode(err, ovirtC.ENotFound) {
```
Comment (janosdebugs) [inline pkg/controllers/providerIDcontroller/providerIDController.go:77]: The other way around.

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]
/lgtm

**Comment by eslutsky** [comment]
the issue preventing the workers to be created is fixed in the last commit.

**Pull request merged**